src/pip/_internal/network/auth.py

```diff
@@ -162,6 +162,49 @@ def _get_new_credentials(
 
         return username, password
 
+    def _get_matching_credentials(self, original_url: str) -> Optional[AuthInfo]:
+        """
+        Find matching credentials based on the longest matching prefix found.
+
+        According to https://datatracker.ietf.org/doc/html/rfc7617#section-2.2
+        the authentication scope is defined by removing the last part after
+        the last "/" of the resource - but in case of `pip` the end of path is always
+        "/project", so we can treat the `original_url` as full `authentication scope'
+        for the request. The RFC specifies that prefix matching of the scope is
+        also within the protection scope specified by the realm value, so we can
+        safely assume that if we find any match, the longest matching prefix is
+        the right authentication information we should use.
+
+        The specification does not decide which of the matching scopes should be
+        used if there are more of them. Our decision is to choose the longest
+        matching one. In case exactly the same prefix will be added several times,
+        the authentication information from the first one is used.
+
+        According to the RFC, the authentication scope includes both scheme and netloc,
+        Both have to match in order to share the authentication scope.
+
+        :param original_url: original URL of the request
+        :return: Stored Authentication info matching the authentication scope or
+                 None if not found
+        """
+        max_matched_prefix_length = 0
+        matched_auth_info = None
+        no_auth_url = remove_auth_from_url(original_url)
+        parsed_original = urllib.parse.urlparse(no_auth_url)
+        for prefix, auth_info in self.passwords:
+            parsed_stored = urllib.parse.urlparse(prefix)
+            if (
+                parsed_stored.netloc != parsed_original.netloc
+                or parsed_stored.scheme != parsed_original.scheme
+            ):
+                # only consider match when both scheme and netloc match
+                continue
```

Comment:
`self.passwords` should probably be a multidict with these as key (something like `Dict[Tuple[netloc, scheme], List[Tuple[prefix, AuthInfo]]]`). The performance of this loop can be a problem pretty quickly.

Reply:
Yeah. I can change it  no problem. It's not really 'likely' to become a problem (you'd have to have a lot of urls with the same netloc to notice the difference - linked list search becomes noticeably slower with quite a few entries. But yeah - why not :).

I will also address the CI failures